src/dense.jl

```diff
@@ -1,193 +1,209 @@
+variable with (Step(t1,y(t1),dy(t1)),Step(t2,y(t2),dy(t2))), where
+`t1` is is the step before `tout` and `t2` is `>=tout`.  In
+other words `tout∈[t1,t2]`.
 
-    # TODO: is this necessary?  The solver should store the value of dy.
-    solver.ode.F!(t0,s0.y,s0.dy)
-    solver.ode.F!(t1,s1.y,s1.dy)
+TODO: tdir
 
-    if s.options.stopevent(t1,s1.y)
-        function stopfun(t)
-            hermite_interp!(state.ytmp,t,s0,s1)
-            res = Int(s.options.stopevent(t,state.ytmp))
-            return 2*res-1      # -1 if false, +1 if true
+"""
+function next_interval!(solver,state,step_prev,tout)
+
+    while true
+        # get the current time
+        t1   = step_prev.t
+        t2,_ = output(state)
+        t1, t2 = sort([t1,t2])
+        if t1 <= tout <= t2
+            # we found the enclosing times
+            return cont
         end
-        t_goal = findroot(stopfun, [s0.t,s1.t], s.options.roottol)
-        # state.ytmp is already overwriten to the correct result as a
-        # side-effect of calling stopfun
-    else
-        hermite_interp!(state.ytmp,t_goal,s0,s1)
-    end
 
-    # update the last output time
-    state.last_tout = t_goal
+        # save the current state of solution
+        t, y, dy = output(state)
+        step_prev.t = t
+        copy!(step_prev.y,y)
+        copy!(step_prev.dy,dy)
 
-    return ((t_goal,state.ytmp),state)
+        # try to perform a single step with the solver
+        status = onestep!(solver.ode, solver.stepper, state)
 
-end
+        if status != cont
+            return status
+        end
+    end
 
+    # this will never happen
+    return abort
+end
 
-function done{O<:ExplicitODE,S<:DenseStepper}(s::Solver{O,S}, state::DenseState)
+"""
 
-    return (
-            state.solver_done ||
-            state.last_tout >= s.options.tspan[end] ||
-            s.options.stopevent(state.s1.t,state.s1.y)
-            )
-end
+Make dense output using Hermite interpolation of order O(3). Updates
+yout in-place.  Only needs y and dy at t1 and t2.
+Input
+- state::AbstractState -- state of a stepper at time t2
+- step_prev::Step -- solution at time t1 respectively
+- tout -- time of requested output
+- yout -- inplace y output
+Ref: Hairer & Wanner p.190
 
+TODO: tdir (I think this works for any order of t1 and t2 but needs
+verifying.
 
-function hermite_interp!(y,t,step0::Step,step1::Step)
-    # For dense output see Hairer & Wanner p.190 using Hermite
-    # interpolation. Updates y in-place.
-    #
-    # f_0 = f(x_0 , y_0) , f_1 = f(x_0 + h, y_1 )
-    # this is O(3). TODO for higher order.
+TODO: fill dy
 
-    y0,  y1  = step0.y, step1.y
-    dy0, dy1 = step0.dy, step1.dy
+TODO: arbitrary order method (change step_prev::Step to step_prevs::Tuple{Step,N})
 
-    if t == step0.t
-        copy!(y,y0)
-    elseif t == step1.t
-        copy!(y,y1)
+"""
+function interpolate!{T,Y}(state::AbstractState,step_prev::Step{T,Y},tout::T,step_out::Step{T,Y})
```

Comment:
I think the signature should be `interpolate!{T,Y}(step_out::Step{T,Y},dstate::DenseState,tout::T)`:

- if possible the in-place variable should be first (although, I'm not really bother by this)
- I think it should take the `dstate`.  Some interpolants may want to store more extra stuff in `DenseState` than just the previous `Step`